Adding IP CIDRs in node annotation

aanm opened the pull request:
- Changed kubernetes client struct to use kubernetes client interface.
- Add functionality to store both IPv4 and IPv6 CIDR in kubernetes node annotations.

tgraf commented on daemon/daemon.go:
```diff
@@ -367,12 +368,36 @@ func (d *Daemon) compileBase() error {
 	return nil
 }
 
+func (d *Daemon) registerK8sAddresses(k8sNode *v1.Node, v4CIDR, v6CIDR *net.IPNet) {
```
We should really move these things to `pkg/k8s` instead of continuing to extend the daemon.

aanm replied:
Fixed!

tgraf commented on pkg/k8s/node.go:
```diff
@@ -63,5 +63,36 @@ func ParseNode(k8sNode *v1.Node) *node.Node {
+	// Spec.PodCIDR takes precedence since it's
+	// the CIDR assigned by k8s controller manager
+	// In case it's invalid or empty then we fall back to our annotations.
+	if node.IPv4AllocCIDR == nil {
```
I think we want to annotation to take precedence. The annotation is more powerful as it indicates IPv4 or IPv6. If the PodCIDR takes precedence and the user wants to use both IPv4 and IPv6, the user has to absolutely ensure that PodCIDR is never set for any reason. That seems error prone. Not annotating seems obvious.

Redo this logic so you can add a:
 ```log.Warningf("k8s: %s annotation with value %s preferred over spec.PodCIDR with value %s", Annotationv4CIDRName, ipv4CIDR, node.IPv4AllocCIDR)```

If both are provided then that is likely a configuration error and we want to report this to the user.

aanm replied:
Ah, I didn't think the user would set the annotations manually, ok on that case it makes sense to switch them precedence order.

What happens if he changes the CIDR after containers have started in the node? The old containers will keep their IP addresses and the new containers will get a new IP addresses, will cilium capable of routing for both CIDRs?

tgraf replied:
Good point. We can't change the CIDR ever. The datapath would actually be ready but the annotations would have to announce all of the CIDRs. Not just a single one.

tgraf commented on pkg/k8s/client.go:
```diff
@@ -89,3 +94,33 @@ func CreateTPRClient(endpoint, kubeCfgPath string) (*rest.RESTClient, error) {
+				// we were able to set the annotations properly
+				k8sServerNode.Annotations[Annotationv4CIDRName] = v4CIDR.String()
+				k8sServerNode.Annotations[Annotationv6CIDRName] = v6CIDR.String()
+				k8sServerNode, err = c.CoreV1().Nodes().Update(k8sNode)
```
Are you sure that simply brute forcing the retry helps? Couldn't the node resource be out of date and you might need to refetch the latest node resource first?

aanm replied:
I haven't test it against k8s but I'm relying on the Update's description:
```
// Update takes the representation of a node and updates it. Returns the server's representation of the node, and an error, if there is any.
```
My understanding is that `k8sServerNode` is the "refetched" node

tgraf replied:
I see, that's good then. I think we should have this running in the background and re-add the annotation in case someone removes the annotation for some reason. We could only trigger the re-add through an agent restart.

tgraf commented on daemon/daemon_test.go:
```diff
@@ -28,3 +37,111 @@ type DaemonSuite struct {
 
 var _ = Suite(&DaemonSuite{})
+
+func (ds *DaemonSuite) TestuseK8sNodeCIDR(c *C) {
```
We can make the k8s client have a default instance like `pkg/kvstore` is done as well. That way, none of this has to be tied to the daemon and we can move all of this code into `pkg/k8s`

tgraf commented on daemon/daemon.go:
```diff
@@ -385,16 +382,35 @@ func (d *Daemon) useK8sNodeCIDR(nodeName string) error {
+		if err != nil {
+			return err
+		}
+		ipv6NodeAddress := nodeaddress.IPv6Address.NodeIP().String()
```
I suggest we wait and rebase this on top of #1020 as this will change heavily and you no longer need to also derive the node address from this.

aanm committed:
k8s: add node IP CIDR in the annotations node

When the user uses the cluster-cidr in the controller manager,
kubernetes controller automatically sets up a dedicated pod CIDR for
each node. This CIDR is used by cilium both locally and on the cluster
so that cilium knows on which node IP is the pod CIDR assigned.

Unfortunately, kubernetes only assigns one IP version CIDR per node, in
consequence, the cilium node derives the other IP version CIDR and
stores this value in the node's attributes map, where the other nodes
receive a notification regarding both IPv4 and IPv6 CIDRs allocated to
that node.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
k8s: change daemon's k8s clientSet to Interface

Changed the daemon's k8s client to use the k8s client interface to
facilitate the use of an interface in future k8s unit tests.

Signed-off-by: André Martins <andre@cilium.io>
